src/UniswapV4MulticurveRehypeInitializerHook.sol

```diff
@@ -0,0 +1,612 @@
+    int128 amount0,
+    int128 amount1,
+    bytes hookData
+);
+
+struct Fees {
+    uint128 fees0;
+    uint128 fees1;
+    uint24 customFee;
+}
+
+uint256 constant MAX_SWAP_FEE = SwapMath.MAX_SWAP_FEE;
+uint256 constant MAX_REBALANCE_ITERATIONS = 15;
+
+// TODO: factor in decimals haha 1e6 maybe for 18 decimals? idk haha
+uint128 constant EPSILON = 1e6;
+
+/**
+ * @title Uniswap V4 Multicurve Hook
+ * @author Whetstone Research
+ * @notice Hook used by the Uniswap V4 Multicurve Initializer to restrict liquidity
+ * addition in a Uniswap V4 pool
+ * @custom:security-contact security@whetstone.cc
+ */
+contract UniswapV4MulticurveRehypeInitializerHook is BaseHook {
+    using StateLibrary for IPoolManager;
+    using ProtocolFeeLibrary for *;
+    using CurrencyLibrary for Currency;
+
+    /// @notice Address of the Uniswap V4 Multicurve Initializer contract
+    address public immutable INITIALIZER;
+    Quoter public immutable quoter;
+
+    struct SwapSimulation {
+        bool success;
+        uint256 amountIn;
+        uint256 amountOut;
+        uint256 fees0;
+        uint256 fees1;
+        uint160 sqrtPriceX96;
+        uint256 excess0;
+        uint256 excess1;
+    }
+
+    mapping(PoolId poolId => Position position) public getPosition;
+    mapping(PoolId poolId => Fees fees) public getFees;
+
+    /**
+     *
+     * @dev Modifier to ensure the caller is the Uniswap V4 Multicurve Initializer
+     * @param sender Address of the caller
+     */
+    modifier onlyInitializer(
+        address sender
+    ) {
+        if (sender != INITIALIZER) revert OnlyInitializer();
+        _;
+    }
+
+    /**
+     * @notice Constructor for the Uniswap V4 Migrator Hook
+     * @param manager Address of the Uniswap V4 Pool Manager
+     * @param initializer Address of the Uniswap V4 Multicurve Initializer contract
+     */
+    constructor(
+        IPoolManager manager,
+        UniswapV4MulticurveInitializer initializer
+    ) BaseHook(manager) {
+        INITIALIZER = address(initializer);
+        quoter = new Quoter(manager);
+    }
+
+    /// @inheritdoc BaseHook
+    function _beforeInitialize(
+        address sender,
+        PoolKey calldata,
+        uint160
+    ) internal view override onlyInitializer(sender) returns (bytes4) {
+        return BaseHook.beforeInitialize.selector;
+    }
+
+    function _afterInitialize(
+        address sender,
+        PoolKey calldata key,
+        uint160,
+        int24
+    ) internal override onlyInitializer(sender) returns (bytes4) {
+        PoolId poolId = key.toId();
+        Position storage position = getPosition[poolId];
+        if (position.salt == bytes32(0)) {
+            position.tickLower = int24(TickMath.minUsableTick(key.tickSpacing));
+            position.tickUpper = int24(TickMath.maxUsableTick(key.tickSpacing));
+            position.salt = _fullRangeSalt(poolId);
+        }
+
+        getFees[poolId] = Fees({ fees0: 0, fees1: 0, customFee: 3000 });
```

Comment:
Note that we currently set a `customFee` of 3000 for every pool so it could be optimized by using a constant instead. But if we expect to make it dynamic later this makes sense